test/qml/tst_editorwidgets.qml

```diff
@@ -2,45 +2,356 @@ import QtQuick 2.3
+    const slider = range.children[1].children[1]
+
+    compare(textField.text, range.default_value + "")
+
+    range.config = {
+      "Style": undefined,
+      "Precision": undefined,
+      "Min": undefined,
+      "Max": undefined,
+      "Step": undefined,
+      "Suffix": undefined
+    }
+    range.value = 3
+    compare(range.widgetStyle, "TextField")
+    compare(range.precision, 1)
+    compare(range.min, -Infinity)
+    compare(range.max, Infinity)
+    compare(range.step, 1)
+    compare(range.suffix, "")
+
+    // Row
+    // compare(sliderRow.visible, true) // ERROR ? should work but not working!
 
-    EditorWidgets.TextEdit {
-        id: textEdit
-        property var value: 'one'
-        property var config: undefined
-        property bool isEnabled: true
+    // TextField -> textField
+    compare(textField.text, "3")
+
+    range.config = {
+      "Style": "Slider",
+      "Precision": 2,
+      "Min": -10,
+      "Max": 10,
+      "Step": 10,
+      "Suffix": "DEFAULT_SUFFIX"
     }
+    range.value = 4
+    compare(range.widgetStyle, "Slider")
+    compare(range.precision, 2)
+    compare(range.min, -10)
+    compare(range.max, 10)
+    compare(range.step, 10)
+    compare(range.suffix, "DEFAULT_SUFFIX")
+
+    compare(sliderRow.visible, false)
+    compare(valueLabel.text, range.min + ".00DEFAULT_SUFFIX") // NOTE: using `range.min` because of `rangeItem.parent.value`
+    compare(slider.value, range.min) // NOTE: using `range.min` because of `rangeItem.parent.value`
+  }
+
+
+  /**
+   * Test case for datetime widget
+   *
+   * This function tests the datetime widget's functionality by:
+   *
+   * - Setting an initial value without setting `config` and `field` objects and verifying that value is empty
+   *
+   * Then, it sets various display formats and verifies that the date and time are displayed correctly in different formats:
+   *
+   * - For each of the test times, it iterates over the display formats and sets the config with the format, calendar popup enabled, unknown field format, and disallows null values
+   * - It sets the field to a date or time type
+   * - It sets the value of the datetime widget to the test time and verifies that the text label is displayed correctly according to the format
+   */
+  function test_01_dateTime() {
+    const label = dateTime.children[1].children[0]
+
+    compare(label.text, "") // NOTE: setting value without setting `config` and `field` objects won't work
+
+    const testTimes = ["2023-01-01", "2023-01-01 23:33:56"]
+    const displayFormats = ["yyyy-MM-dd", "yyyy-MM.dd", "yyyy-MM-dd HH:mm:ss", "HH:mm:ss", "HH:mm"]
+    const results = ["2023-01-01", "2023-01.01", "2023-01-01 00:00:00", "00:00:00", "00:00",
+                     "2023-01-01", "2023-01.01", "2023-01-01 23:33:56", "23:33:56", "23:33"]
+    let resultIdx = 0
+
+    for (let time of testTimes) {
+      for (let format of displayFormats) {
+        dateTime.config = {
```

Comment:
Nice, I love how you're stress-testing values and formats.

Reply:
thanks